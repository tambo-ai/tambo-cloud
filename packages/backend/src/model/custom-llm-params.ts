/**
 * Custom LLM Parameters Types and Utilities
 *
 * This module provides types and utilities for handling custom LLM parameters
 * that can be passed to providers with proper precedence and validation.
 */

/**
 * Custom LLM parameters that can be passed to providers.
 * Kept permissive to allow provider-specific parameters.
 */
export type CustomLlmParams = Record<string, unknown>;

/**
 * Reserved/locked keys that should not be overridden by custom parameters.
 * These keys are critical for routing, observability, and core functionality.
 */
export const RESERVED_LLM_PARAM_KEYS = new Set([
  "model",
  "messages",
  "tools",
  "tool_choice",
  "response_format",
  "stream",
  "user",
  "logit_bias",
  "logprobs",
  "top_logprobs",
  "n",
  "parallel_tool_calls",
  "service_tier",
  "stop",
  "stream_options",
  "metadata",
  "store",
]);

/**
 * Standard LLM parameters that should be passed as normal model options.
 * These are common parameters supported across most providers.
 */
export const STANDARD_LLM_PARAM_KEYS = new Set([
  "temperature",
  "top_p",
  "max_tokens",
  "max_completion_tokens",
  "presence_penalty",
  "frequency_penalty",
  "seed",
]);

/**
 * Parameters for the mergeModelParams function.
 */
export interface MergeModelParamsInput {
  /** Default parameters (lowest precedence) */
  defaults?: CustomLlmParams;
  /** Project-level parameters (medium precedence) */
  projectParams?: CustomLlmParams;
  /** Request-level overrides (highest precedence) */
  requestOverrides?: CustomLlmParams;
}

/**
 * Result of merging model parameters.
 */
export interface MergedModelParams {
  /** Standard parameters that should be passed as normal model options */
  standardParams: Record<string, unknown>;
  /** Provider-specific parameters that should be passed via providerOptions */
  providerOptions: Record<string, unknown>;
  /** All merged parameters (for debugging/logging) */
  allParams: CustomLlmParams;
}

/**
 * Deep merge utility for objects.
 * Later objects override earlier ones, with deep merging for nested objects.
 */
function deepMerge(
  target: Record<string, unknown>,
  source: Record<string, unknown>,
): Record<string, unknown> {
  const result = { ...target };

  for (const [key, value] of Object.entries(source)) {
    if (value === null || value === undefined) {
      result[key] = value;
    } else if (
      typeof value === "object" &&
      !Array.isArray(value) &&
      typeof result[key] === "object" &&
      !Array.isArray(result[key]) &&
      result[key] !== null
    ) {
      result[key] = deepMerge(
        result[key] as Record<string, unknown>,
        value as Record<string, unknown>,
      );
    } else {
      result[key] = value;
    }
  }

  return result;
}

/**
 * Strips reserved/locked keys from parameters.
 * Logs warnings for any stripped keys.
 */
function stripReservedKeys(
  params: CustomLlmParams,
  logger?: { debug: (message: string) => void },
): CustomLlmParams {
  const result: CustomLlmParams = {};
  const strippedKeys: string[] = [];

  for (const [key, value] of Object.entries(params)) {
    if (RESERVED_LLM_PARAM_KEYS.has(key)) {
      strippedKeys.push(key);
    } else {
      result[key] = value;
    }
  }

  if (strippedKeys.length > 0 && logger) {
    logger.debug(
      `Stripped reserved LLM parameter keys: ${strippedKeys.join(", ")}`,
    );
  }

  return result;
}

/**
 * Merges model parameters with proper precedence: defaults < projectParams < requestOverrides.
 * Performs deep merging for nested objects and strips reserved keys.
 *
 * @param input - The parameters to merge
 * @param logger - Optional logger for debug messages
 * @returns Merged parameters split into standard params and provider options
 */
export function mergeModelParams(
  input: MergeModelParamsInput,
  logger?: { debug: (message: string) => void },
): MergedModelParams {
  const { defaults = {}, projectParams = {}, requestOverrides = {} } = input;

  // Merge with proper precedence (last-in wins)
  let merged = deepMerge({}, defaults);
  merged = deepMerge(merged, projectParams);
  merged = deepMerge(merged, requestOverrides);

  // Strip reserved keys
  const cleanParams = stripReservedKeys(merged, logger);

  // Split into standard params and provider options
  const standardParams: Record<string, unknown> = {};
  const providerOptions: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(cleanParams)) {
    if (STANDARD_LLM_PARAM_KEYS.has(key)) {
      standardParams[key] = value;
    } else {
      providerOptions[key] = value;
    }
  }

  if (logger && Object.keys(providerOptions).length > 0) {
    logger.debug(
      `Passing provider-specific options: ${Object.keys(providerOptions).join(", ")}`,
    );
  }

  return {
    standardParams,
    providerOptions,
    allParams: cleanParams,
  };
}

/**
 * Validates that custom LLM parameters are a valid JSON object.
 *
 * @param params - The parameters to validate
 * @returns True if valid, false otherwise
 */
export function validateCustomLlmParams(
  params: unknown,
): params is CustomLlmParams {
  return (
    typeof params === "object" && params !== null && !Array.isArray(params)
  );
}

/**
 * Gets the size of a JSON object in bytes (approximate).
 * Used for validating parameter payload size limits.
 *
 * @param params - The parameters to measure
 * @returns Size in bytes
 */
export function getCustomLlmParamsSize(params: CustomLlmParams): number {
  return new TextEncoder().encode(JSON.stringify(params)).length;
}

/**
 * Maximum allowed size for custom LLM parameters (16KB).
 */
export const MAX_CUSTOM_LLM_PARAMS_SIZE = 16 * 1024; // 16KB
