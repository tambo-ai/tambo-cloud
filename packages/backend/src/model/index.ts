export * from "./component-metadata";
export * from "./providers";
