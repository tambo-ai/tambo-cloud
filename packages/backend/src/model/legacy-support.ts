import {
  ComponentContextToolMetadata,
  ToolResponseBody,
} from "./component-metadata";

export interface ComponentContextTool<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Args extends unknown[] = any[],
  Response extends ToolResponseBody = ToolResponseBody,
> {
  getComponentContext: (...args: Args) => Promise<Response>;
  definition: ComponentContextToolMetadata;
}
