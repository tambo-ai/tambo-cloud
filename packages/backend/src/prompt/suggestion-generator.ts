import { MessageRole, ThreadMessage } from "@tambo-ai-cloud/core";
import {
  AvailableComponent,
  AvailableComponents,
} from "../model/component-metadata";
import { getAvailableComponentsPromptTemplate } from "./component-formatting";

/**
 * @param components List of available components
 * @param messages Array of thread messages to provide context
 * @param suggestionCount Number of suggestions to generate
 * @returns Array of system and user messages
 */
export function buildSuggestionPrompt(
  components: AvailableComponent[],
  messages: ThreadMessage[] = [],
  suggestionCount: number,
): Array<{ role: MessageRole; content: string }> {
  // Get current component if available
  let currentComponent: string | null = null;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let currentProps: any | null = null;

  if (messages.length > 0) {
    for (let i = messages.length - 1; i >= 0; i--) {
      const msg = messages[i];
      if (msg.component?.componentName) {
        currentComponent = msg.component.componentName;
        currentProps = msg.component.props || null;
        break;
      }
    }
  }

  // Convert components array to AvailableComponents object format
  const availableComponentsObj: AvailableComponents = {};
  components.forEach((component) => {
    availableComponentsObj[component.name] = component;
  });

  // Use getAvailableComponentsPromptTemplate to generate the component list
  const availableComponentsTemplate = getAvailableComponentsPromptTemplate(
    availableComponentsObj,
  );
  const componentList = availableComponentsTemplate.template
    .replace(
      "{availableComponents}",
      availableComponentsTemplate.args.availableComponents,
    )
    .trim();

  const systemMessage = `Review the available components and conversation history to generate natural follow-up messages that a user might send.

${componentList}

Your task is to suggest ${suggestionCount} messages written exactly as if they came from the user. These suggestions should represent natural follow-up requests based on the available components and context.

Rules:
0. Make absolutely sure to respond in the same language as the user.
1. Write each suggestion as a complete message that could be sent by the user
2. Focus on practical requests that use the available components
3. Make suggestions contextually relevant to the conversation and previous actions
4. If a component is currently in use, suggest natural variations or new ways to use it
5. Write in a natural, conversational tone as if the user is typing
6. Avoid technical language or system-focused phrasing`;

  // Build the result array starting with the system message
  const result: Array<{
    role: MessageRole;
    content: string;
  }> = [{ role: MessageRole.System, content: systemMessage }];

  // Add recent conversation history as separate messages
  if (messages.length > 0) {
    const recentMessages = messages.slice(-2);
    recentMessages.forEach((msg) => {
      const content = msg.content
        .map((c) => ("text" in c ? c.text : ""))
        .join("");
      if (content.trim()) {
        result.push({ role: msg.role, content });
      }
    });
  }

  // Add current component context if available
  const componentContext = currentComponent
    ? `Current component: ${currentComponent}\nCurrent props: ${JSON.stringify(currentProps, null, 2)}`
    : "No component currently in use.";

  // Add the final user message with the request
  const finalUserMessage = `Generate ${suggestionCount} natural follow-up messages that a user might send. Each suggestion should be a complete message that could be sent directly to the system.

The suggestions should be written exactly as a user would type them, not as descriptions or commands, in a JSON structure.

The suggestions should be written in the same language as the above conversation.

${componentContext}`;

  result.push({ role: MessageRole.User, content: finalUserMessage });

  return result;
}
