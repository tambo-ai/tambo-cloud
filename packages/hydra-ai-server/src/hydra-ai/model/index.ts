export * from "./chat-message";
export * from "./component-metadata";
export * from "./component-props-metadata";
export * from "./input-context";
export * from "./providers";
