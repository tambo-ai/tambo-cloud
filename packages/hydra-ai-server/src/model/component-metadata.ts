import { type JSONSchema7 } from "json-schema";
import { ComponentPropsMetadata } from "./component-props-metadata";

export interface ComponentMetadata {
  name: string;
  description: string;
  props: ComponentPropsMetadata;
}

export interface ComponentContextToolMetadata {
  name: string;
  description: string;
  parameters: {
    name: string;
    type: string;
    description: string;
    isRequired: boolean;
    items?: { type: string };
    enumValues?: string[];
    schema?: JSONSchema7;
  }[];
}

export interface ComponentContextTool<
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Args extends unknown[] = any[],
  Response extends ToolResponseBody = ToolResponseBody,
> {
  getComponentContext: (...args: Args) => Promise<Response>;
  definition: ComponentContextToolMetadata;
}

export interface AvailableComponent extends ComponentMetadata {
  contextTools: ComponentContextToolMetadata[];
}

export interface ComponentWithContext extends ComponentMetadata {
  context: ToolResponseBody;
}

export interface AvailableComponents {
  [key: string]: AvailableComponent;
}

/**
 * The body of a tool response.
 * This is the data returned by the tool.
 * This is not the tool response, which is the entire message from the tool.
 */
export type ToolResponseBody = unknown;
