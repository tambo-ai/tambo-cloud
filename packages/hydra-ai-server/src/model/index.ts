export * from "./chat-message";
export * from "./component-metadata";
export * from "./input-context";
export * from "./legacy-support";
export * from "./openai-response";
export * from "./providers";
