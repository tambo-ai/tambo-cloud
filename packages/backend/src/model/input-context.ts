import { ThreadMessage } from "@tambo-ai-cloud/core";
import { AvailableComponent } from "./component-metadata";

export type InputContextAsArray = {
  messageHistory: ThreadMessage[];
  availableComponents: AvailableComponent[];
  threadId: string;
};
