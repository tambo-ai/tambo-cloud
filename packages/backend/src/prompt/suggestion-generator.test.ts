import {
  LegacyComponentDecision,
  MessageRole,
  ThreadMessage,
} from "@tambo-ai-cloud/core";
import { AvailableComponent } from "../model/component-metadata";
import { buildSuggestionPrompt } from "./suggestion-generator";

describe("suggestion-generator", () => {
  const mockComponents: AvailableComponent[] = [
    {
      name: "Button",
      description: "A simple button component",
      props: {
        label: "string",
        onClick: "(event: MouseEvent) => void",
      },
      contextTools: [],
    },
    {
      name: "Input",
      description: "An input field component",
      props: {
        value: {
          type: "string",
          description: "Current input value",
          required: true,
        },
        onChange: {
          type: "(value: string) => void",
          description: "Change handler",
          required: true,
        },
      },
      contextTools: [],
    },
  ];

  describe("buildSuggestionPrompt", () => {
    it("should generate prompt with no conversation history", () => {
      const result = buildSuggestionPrompt(mockComponents, [], 3);

      expect(result).toHaveLength(2);
      expect(result[0].role).toBe("system");
      expect(result[1].role).toBe("user");

      expect(result).toMatchInlineSnapshot(`
        [
          {
            "content": "Review the available components and conversation history to generate natural follow-up messages that a user might send.

        You may use only the following components:

        - componentName: "Button":
            description: A simple button component
            props:
            {
              "label": "string",
              "onClick": "(event: MouseEvent) => void"
            }

        - componentName: "Input":
            description: An input field component
            props:
            {
              "value": {
                "type": "string",
                "description": "Current input value",
                "required": true
              },
              "onChange": {
                "type": "(value: string) => void",
                "description": "Change handler",
                "required": true
              }
            }

        Your task is to suggest 3 messages written exactly as if they came from the user. These suggestions should represent natural follow-up requests based on the available components and context.

        Rules:
        0. Make absolutely sure to respond in the same language as the user.
        1. Write each suggestion as a complete message that could be sent by the user
        2. Focus on practical requests that use the available components
        3. Make suggestions contextually relevant to the conversation and previous actions
        4. If a component is currently in use, suggest natural variations or new ways to use it
        5. Write in a natural, conversational tone as if the user is typing
        6. Avoid technical language or system-focused phrasing",
            "role": "system",
          },
          {
            "content": "Generate 3 natural follow-up messages that a user might send. Each suggestion should be a complete message that could be sent directly to the system.

        The suggestions should be written exactly as a user would type them, not as descriptions or commands, in a JSON structure.

        The suggestions should be written in the same language as the above conversation.

        No component currently in use.",
            "role": "user",
          },
        ]
      `);
    });

    it("should include conversation history when available", () => {
      const mockHistory: ThreadMessage[] = [
        {
          role: MessageRole.User,
          content: [{ type: "text", text: "Show me a button" }],
          id: "1",
          threadId: "thread-1",
          componentState: {},
          createdAt: new Date("2024-01-01T00:00:00.000Z"), // Fixed date for snapshot
        },
        {
          role: MessageRole.Assistant,
          content: [{ type: "text", text: "Here's your button" }],
          id: "2",
          threadId: "thread-1",
          componentState: {},
          createdAt: new Date("2024-01-01T00:00:01.000Z"), // Fixed date for snapshot
        },
      ];

      const result = buildSuggestionPrompt(mockComponents, mockHistory, 2);

      expect(result).toMatchInlineSnapshot(`
        [
          {
            "content": "Review the available components and conversation history to generate natural follow-up messages that a user might send.

        You may use only the following components:

        - componentName: "Button":
            description: A simple button component
            props:
            {
              "label": "string",
              "onClick": "(event: MouseEvent) => void"
            }

        - componentName: "Input":
            description: An input field component
            props:
            {
              "value": {
                "type": "string",
                "description": "Current input value",
                "required": true
              },
              "onChange": {
                "type": "(value: string) => void",
                "description": "Change handler",
                "required": true
              }
            }

        Your task is to suggest 2 messages written exactly as if they came from the user. These suggestions should represent natural follow-up requests based on the available components and context.

        Rules:
        0. Make absolutely sure to respond in the same language as the user.
        1. Write each suggestion as a complete message that could be sent by the user
        2. Focus on practical requests that use the available components
        3. Make suggestions contextually relevant to the conversation and previous actions
        4. If a component is currently in use, suggest natural variations or new ways to use it
        5. Write in a natural, conversational tone as if the user is typing
        6. Avoid technical language or system-focused phrasing",
            "role": "system",
          },
          {
            "content": "Show me a button",
            "role": "user",
          },
          {
            "content": "Here's your button",
            "role": "assistant",
          },
          {
            "content": "Generate 2 natural follow-up messages that a user might send. Each suggestion should be a complete message that could be sent directly to the system.

        The suggestions should be written exactly as a user would type them, not as descriptions or commands, in a JSON structure.

        The suggestions should be written in the same language as the above conversation.

        No component currently in use.",
            "role": "user",
          },
        ]
      `);
    });

    it("should handle current component context", () => {
      const componentDecision: LegacyComponentDecision = {
        componentName: "Button",
        props: { label: "Click me" },
        message: "Using button component",
        componentState: {},
        reasoning: "Button requested by user",
      };

      const mockHistory: ThreadMessage[] = [
        {
          role: MessageRole.User,
          content: [{ type: "text", text: "Show me a button" }],
          id: "1",
          threadId: "thread-1",
          componentState: {},
          createdAt: new Date("2024-01-01T00:00:00.000Z"), // Fixed date for snapshot
          component: componentDecision,
        },
      ];

      const result = buildSuggestionPrompt(mockComponents, mockHistory, 2);

      expect(result).toMatchInlineSnapshot(`
        [
          {
            "content": "Review the available components and conversation history to generate natural follow-up messages that a user might send.

        You may use only the following components:

        - componentName: "Button":
            description: A simple button component
            props:
            {
              "label": "string",
              "onClick": "(event: MouseEvent) => void"
            }

        - componentName: "Input":
            description: An input field component
            props:
            {
              "value": {
                "type": "string",
                "description": "Current input value",
                "required": true
              },
              "onChange": {
                "type": "(value: string) => void",
                "description": "Change handler",
                "required": true
              }
            }

        Your task is to suggest 2 messages written exactly as if they came from the user. These suggestions should represent natural follow-up requests based on the available components and context.

        Rules:
        0. Make absolutely sure to respond in the same language as the user.
        1. Write each suggestion as a complete message that could be sent by the user
        2. Focus on practical requests that use the available components
        3. Make suggestions contextually relevant to the conversation and previous actions
        4. If a component is currently in use, suggest natural variations or new ways to use it
        5. Write in a natural, conversational tone as if the user is typing
        6. Avoid technical language or system-focused phrasing",
            "role": "system",
          },
          {
            "content": "Show me a button",
            "role": "user",
          },
          {
            "content": "Generate 2 natural follow-up messages that a user might send. Each suggestion should be a complete message that could be sent directly to the system.

        The suggestions should be written exactly as a user would type them, not as descriptions or commands, in a JSON structure.

        The suggestions should be written in the same language as the above conversation.

        Current component: Button
        Current props: {
          "label": "Click me"
        }",
            "role": "user",
          },
        ]
      `);
    });

    it("should handle empty component list", () => {
      const result = buildSuggestionPrompt([], [], 1);

      expect(result).toMatchInlineSnapshot(`
        [
          {
            "content": "Review the available components and conversation history to generate natural follow-up messages that a user might send.

        You may use only the following components:
        No components available, do not try and generate a component.

        Your task is to suggest 1 messages written exactly as if they came from the user. These suggestions should represent natural follow-up requests based on the available components and context.

        Rules:
        0. Make absolutely sure to respond in the same language as the user.
        1. Write each suggestion as a complete message that could be sent by the user
        2. Focus on practical requests that use the available components
        3. Make suggestions contextually relevant to the conversation and previous actions
        4. If a component is currently in use, suggest natural variations or new ways to use it
        5. Write in a natural, conversational tone as if the user is typing
        6. Avoid technical language or system-focused phrasing",
            "role": "system",
          },
          {
            "content": "Generate 1 natural follow-up messages that a user might send. Each suggestion should be a complete message that could be sent directly to the system.

        The suggestions should be written exactly as a user would type them, not as descriptions or commands, in a JSON structure.

        The suggestions should be written in the same language as the above conversation.

        No component currently in use.",
            "role": "user",
          },
        ]
      `);
    });

    it("should include correct number of suggestions in prompt", () => {
      const counts = [1, 3, 5];

      counts.forEach((count) => {
        const result = buildSuggestionPrompt(mockComponents, [], count);
        expect(result).toMatchSnapshot();
      });
    });

    it("should handle component with complex props", () => {
      const complexComponent: AvailableComponent[] = [
        {
          name: "ComplexForm",
          description: "A form with nested props",
          props: {
            formData: {
              type: "object",
              description: "Form data object",
              required: true,
            },
            validation: {
              type: "{ rules: object; messages: object }",
              description: "Validation configuration",
            },
          },
          contextTools: [],
        },
      ];

      const result = buildSuggestionPrompt(complexComponent, [], 1);

      expect(result).toMatchInlineSnapshot(`
        [
          {
            "content": "Review the available components and conversation history to generate natural follow-up messages that a user might send.

        You may use only the following components:

        - componentName: "ComplexForm":
            description: A form with nested props
            props:
            {
              "formData": {
                "type": "object",
                "description": "Form data object",
                "required": true
              },
              "validation": {
                "type": "{ rules: object; messages: object }",
                "description": "Validation configuration"
              }
            }

        Your task is to suggest 1 messages written exactly as if they came from the user. These suggestions should represent natural follow-up requests based on the available components and context.

        Rules:
        0. Make absolutely sure to respond in the same language as the user.
        1. Write each suggestion as a complete message that could be sent by the user
        2. Focus on practical requests that use the available components
        3. Make suggestions contextually relevant to the conversation and previous actions
        4. If a component is currently in use, suggest natural variations or new ways to use it
        5. Write in a natural, conversational tone as if the user is typing
        6. Avoid technical language or system-focused phrasing",
            "role": "system",
          },
          {
            "content": "Generate 1 natural follow-up messages that a user might send. Each suggestion should be a complete message that could be sent directly to the system.

        The suggestions should be written exactly as a user would type them, not as descriptions or commands, in a JSON structure.

        The suggestions should be written in the same language as the above conversation.

        No component currently in use.",
            "role": "user",
          },
        ]
      `);
    });

    it("should preserve most recent component context", () => {
      const buttonDecision: LegacyComponentDecision = {
        componentName: "Button",
        props: { label: "First" },
        message: "Using button component",
        componentState: {},
        reasoning: "Button requested by user",
      };

      const inputDecision: LegacyComponentDecision = {
        componentName: "Input",
        props: { value: "Latest" },
        message: "Using input component",
        componentState: {},
        reasoning: "Input requested by user",
      };

      const mockHistory: ThreadMessage[] = [
        {
          role: MessageRole.User,
          content: [{ type: "text", text: "Show button" }],
          id: "1",
          threadId: "thread-1",
          componentState: {},
          createdAt: new Date("2024-01-01T00:00:00.000Z"), // Fixed date for snapshot
          component: buttonDecision,
        },
        {
          role: MessageRole.User,
          content: [{ type: "text", text: "Show input" }],
          id: "2",
          threadId: "thread-1",
          componentState: {},
          createdAt: new Date("2024-01-01T00:00:01.000Z"), // Fixed date for snapshot
          component: inputDecision,
        },
      ];

      const result = buildSuggestionPrompt(mockComponents, mockHistory, 1);

      expect(result).toMatchInlineSnapshot(`
        [
          {
            "content": "Review the available components and conversation history to generate natural follow-up messages that a user might send.

        You may use only the following components:

        - componentName: "Button":
            description: A simple button component
            props:
            {
              "label": "string",
              "onClick": "(event: MouseEvent) => void"
            }

        - componentName: "Input":
            description: An input field component
            props:
            {
              "value": {
                "type": "string",
                "description": "Current input value",
                "required": true
              },
              "onChange": {
                "type": "(value: string) => void",
                "description": "Change handler",
                "required": true
              }
            }

        Your task is to suggest 1 messages written exactly as if they came from the user. These suggestions should represent natural follow-up requests based on the available components and context.

        Rules:
        0. Make absolutely sure to respond in the same language as the user.
        1. Write each suggestion as a complete message that could be sent by the user
        2. Focus on practical requests that use the available components
        3. Make suggestions contextually relevant to the conversation and previous actions
        4. If a component is currently in use, suggest natural variations or new ways to use it
        5. Write in a natural, conversational tone as if the user is typing
        6. Avoid technical language or system-focused phrasing",
            "role": "system",
          },
          {
            "content": "Show button",
            "role": "user",
          },
          {
            "content": "Show input",
            "role": "user",
          },
          {
            "content": "Generate 1 natural follow-up messages that a user might send. Each suggestion should be a complete message that could be sent directly to the system.

        The suggestions should be written exactly as a user would type them, not as descriptions or commands, in a JSON structure.

        The suggestions should be written in the same language as the above conversation.

        Current component: Input
        Current props: {
          "value": "Latest"
        }",
            "role": "user",
          },
        ]
      `);
    });
  });
});
