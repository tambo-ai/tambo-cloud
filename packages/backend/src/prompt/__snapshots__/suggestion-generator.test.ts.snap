// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`suggestion-generator buildSuggestionPrompt should include correct number of suggestions in prompt 1`] = `
[
  {
    "content": "Review the available components and conversation history to generate natural follow-up messages that a user might send.

You may use only the following components:

- componentName: "Button":
    description: A simple button component
    props:
    {
      "label": "string",
      "onClick": "(event: MouseEvent) => void"
    }

- componentName: "Input":
    description: An input field component
    props:
    {
      "value": {
        "type": "string",
        "description": "Current input value",
        "required": true
      },
      "onChange": {
        "type": "(value: string) => void",
        "description": "Change handler",
        "required": true
      }
    }

Your task is to suggest 1 messages written exactly as if they came from the user. These suggestions should represent natural follow-up requests based on the available components and context.

Rules:
0. Make absolutely sure to respond in the same language as the user.
1. Write each suggestion as a complete message that could be sent by the user
2. Focus on practical requests that use the available components
3. Make suggestions contextually relevant to the conversation and previous actions
4. If a component is currently in use, suggest natural variations or new ways to use it
5. Write in a natural, conversational tone as if the user is typing
6. Avoid technical language or system-focused phrasing",
    "role": "system",
  },
  {
    "content": "No conversation history yet.

Generate 1 natural follow-up messages that a user might send. Each suggestion should be a complete message that could be sent directly to the system.

The suggestions should be written exactly as a user would type them, not as descriptions or commands, in a JSON structure.
",
    "role": "user",
  },
]
`;

exports[`suggestion-generator buildSuggestionPrompt should include correct number of suggestions in prompt 2`] = `
[
  {
    "content": "Review the available components and conversation history to generate natural follow-up messages that a user might send.

You may use only the following components:

- componentName: "Button":
    description: A simple button component
    props:
    {
      "label": "string",
      "onClick": "(event: MouseEvent) => void"
    }

- componentName: "Input":
    description: An input field component
    props:
    {
      "value": {
        "type": "string",
        "description": "Current input value",
        "required": true
      },
      "onChange": {
        "type": "(value: string) => void",
        "description": "Change handler",
        "required": true
      }
    }

Your task is to suggest 3 messages written exactly as if they came from the user. These suggestions should represent natural follow-up requests based on the available components and context.

Rules:
0. Make absolutely sure to respond in the same language as the user.
1. Write each suggestion as a complete message that could be sent by the user
2. Focus on practical requests that use the available components
3. Make suggestions contextually relevant to the conversation and previous actions
4. If a component is currently in use, suggest natural variations or new ways to use it
5. Write in a natural, conversational tone as if the user is typing
6. Avoid technical language or system-focused phrasing",
    "role": "system",
  },
  {
    "content": "No conversation history yet.

Generate 3 natural follow-up messages that a user might send. Each suggestion should be a complete message that could be sent directly to the system.

The suggestions should be written exactly as a user would type them, not as descriptions or commands, in a JSON structure.
",
    "role": "user",
  },
]
`;

exports[`suggestion-generator buildSuggestionPrompt should include correct number of suggestions in prompt 3`] = `
[
  {
    "content": "Review the available components and conversation history to generate natural follow-up messages that a user might send.

You may use only the following components:

- componentName: "Button":
    description: A simple button component
    props:
    {
      "label": "string",
      "onClick": "(event: MouseEvent) => void"
    }

- componentName: "Input":
    description: An input field component
    props:
    {
      "value": {
        "type": "string",
        "description": "Current input value",
        "required": true
      },
      "onChange": {
        "type": "(value: string) => void",
        "description": "Change handler",
        "required": true
      }
    }

Your task is to suggest 5 messages written exactly as if they came from the user. These suggestions should represent natural follow-up requests based on the available components and context.

Rules:
0. Make absolutely sure to respond in the same language as the user.
1. Write each suggestion as a complete message that could be sent by the user
2. Focus on practical requests that use the available components
3. Make suggestions contextually relevant to the conversation and previous actions
4. If a component is currently in use, suggest natural variations or new ways to use it
5. Write in a natural, conversational tone as if the user is typing
6. Avoid technical language or system-focused phrasing",
    "role": "system",
  },
  {
    "content": "No conversation history yet.

Generate 5 natural follow-up messages that a user might send. Each suggestion should be a complete message that could be sent directly to the system.

The suggestions should be written exactly as a user would type them, not as descriptions or commands, in a JSON structure.
",
    "role": "user",
  },
]
`;
